pageserver/src/tenant.rs

```diff
@@ -2647,36 +2647,46 @@ pub mod harness {
         }
     }
 
-    pub struct TenantHarness<'a> {
+    /// The harness provides a way to create mock certain tenant processes for testing.
+    /// It also creates a temporary dir that serves as a test repository, that gets removed on the harness drop.
+    pub struct TenantHarness {
+        // keep the struct to not to remove tmp dir during the test
+        temp_repo_dir: Option<TempDir>,
         pub conf: &'static PageServerConf,
         pub tenant_conf: TenantConf,
         pub tenant_id: TenantId,
-
-        pub lock_guard: (
-            Option<RwLockReadGuard<'a, ()>>,
-            Option<RwLockWriteGuard<'a, ()>>,
-        ),
     }
 
-    impl<'a> TenantHarness<'a> {
-        pub fn create(test_name: &'static str) -> anyhow::Result<Self> {
-            Self::create_internal(test_name, false)
-        }
-        pub fn create_exclusive(test_name: &'static str) -> anyhow::Result<Self> {
-            Self::create_internal(test_name, true)
-        }
-        fn create_internal(test_name: &'static str, exclusive: bool) -> anyhow::Result<Self> {
-            let lock_guard = if exclusive {
-                (None, Some(LOCK.write().unwrap()))
+    impl Drop for TenantHarness {
+        fn drop(&mut self) {
+            if let Some(temp_repo_dir) = self.temp_repo_dir.take() {
+                temp_repo_dir.close().unwrap_or_else(|e| {
+                    panic!(
+                        "Test tmp directory {:?} failed to be removed: {}",
+                        self.temp_repo_dir, e
+                    )
+                })
             } else {
-                (Some(LOCK.read().unwrap()), None)
-            };
+                panic!("No temporary dir found on harness drop")
+            }
+        }
+    }
+
+    static LOG_HANDLE: OnceCell<()> = OnceCell::new();
+
+    impl TenantHarness {
+        pub fn new() -> anyhow::Result<Self> {
+            LOG_HANDLE.get_or_init(|| {
+                logging::init(logging::LogFormat::Test).expect("Failed to init test logging")
+            });
 
-            let repo_dir = PageServerConf::test_repo_dir(test_name);
-            let _ = fs::remove_dir_all(&repo_dir);
-            fs::create_dir_all(&repo_dir)?;
+            let temp_repo_dir = tempfile::tempdir()?;
+            // `TempDir` uses a randomly generated subdirectory of a system tmp dir,
+            // so far it's enough to take care of concurrently running tests.
+            let repo_dir = temp_repo_dir.path();
+            fs::create_dir_all(repo_dir)?;
```

Comment:
`create_dir_all` looks redundant since tempdir actually creates it, not just generates a path